### PR TITLE
add open device registration test with cpp extensions

### PR DESCRIPTION
**Pull request by bdhirsh**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #80477
* #80418
* #80416

Adding a test for open device registration using cpp extensions.

### CODE REVIEW

File: test/test_cpp_extensions_open_device_registration.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    CUDNN_HEADER_EXISTS = os.path.isfile(os.path.join(CUDA_HOME, "include/cudnn.h"))
+    TEST_CUDNN = (
+        TEST_CUDA and CUDNN_HEADER_EXISTS and torch.backends.cudnn.is_available()
+    )
+IS_WINDOWS = sys.platform == "win32"
+
+
+def remove_build_path():
+    if sys.platform == "win32":
+        print("Not wiping extensions build folder because Windows")
```
Comment (albanD) [inline test/test_cpp_extensions_open_device_registration.py:26]: print?

File: test/test_cpp_extensions_open_device_registration.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        self.assertFalse(module.custom_add_called())
+
+        # create a tensor using our custom device object.
+        device = module.custom_device()
+        x = torch.empty(4, 4, device=device)
+        y = torch.empty(4, 4, device=device)
+
+        self.assertFalse(module.custom_add_called())
+
+        # calls out custom add kernel, registered to the dispatcher
```
Comment (albanD) [inline test/test_cpp_extensions_open_device_registration.py:77]: nit: our

File: test/cpp_extensions/open_registration_extension.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  TORCH_CHECK(self.sizes() == dst.sizes());
+  TORCH_CHECK(self.scalar_type() == dst.scalar_type());
+  TORCH_CHECK(self.is_contiguous() && dst.is_contiguous());
+
+  std::memcpy(dst.storage().data_ptr().get(), self.storage().data_ptr().get(), self.storage().nbytes());
+  return dst;
+}
+
+
+TORCH_LIBRARY_IMPL(aten, PrivateUse1, m) {
```
Comment (albanD) [inline test/cpp_extensions/open_registration_extension.cpp:73]: nit: add a comment that explains what this is doing?

File: test/cpp_extensions/open_registration_extension.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  m.impl("add.Tensor", &custom_add_Tensor);
+  m.impl("empty.memory_format", &custom_empty_memory_format);
+  m.impl("fill_.Scalar", &custom_fill__scalar);
+  m.impl("_copy_from", &custom__copy_from);
+}
+
+c10::Device get_custom_device() {
```
Comment (albanD) [inline test/cpp_extensions/open_registration_extension.cpp:80]: nit why is this needed? Can it be done directly in python?

File: test/cpp_extensions/open_registration_extension.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+bool custom_add_called() {
+  bool called = false;
+  if (add_counter > last_saved_value) {
+    called = true;
+    last_saved_value = add_counter;
+  }
+  return called;
+}
+
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
```
Comment (albanD) [inline test/cpp_extensions/open_registration_extension.cpp:93]: nit: comment explaining why we need this.

File: test/test_cpp_extensions_open_device_registration.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        z = x + y
+
+        # check that it was called
+        self.assertTrue(module.custom_add_called())
+
+        z_cpu = z.to(device='cpu')
```
Comment (albanD) [inline test/test_cpp_extensions_open_device_registration.py:83]: nit: test that the "cross device" copy did copy the values properly?

**Review by albanD — COMMENTED** [review]
Looks great!
I think the second part of the c++ file could use similar comments to the first part of it. Added comments inline to that end.

File: test/cpp_extensions/open_registration_extension.cpp (hunk 3 above)
Comment (bdhirsh) [reply test/cpp_extensions/open_registration_extension.cpp:73]: agreed

File: test/test_cpp_extensions_open_device_registration.py (hunk 1 above)
Comment (bdhirsh) [reply test/test_cpp_extensions_open_device_registration.py:26]: I copied this from an existing cpp extension test we have, but agreed - will clean up

File: test/cpp_extensions/open_registration_extension.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+// For open registration, we're registering all of our kernels to the PrivateUse1 dispatch key.
+// Later in this file, we map a custom device to the PrivateUse1 device type,
+// which allows user code that puts a tensor on your custom_device to eventually get plumbed
+// into the kernels registered here.
+//
+// This macro registers your kernels to the PyTorch Dispatcher.
+// More details on the dispatcher can be found at http://blog.ezyang.com/2020/09/lets-talk-about-the-pytorch-dispatcher/.
+TORCH_LIBRARY_IMPL(aten, PrivateUse1, m) {
+  m.impl("add.Tensor", &custom_add_Tensor);
+  m.impl("empty.memory_format", &custom_empty_memory_format);
```
Comment (albanD) [inline test/cpp_extensions/open_registration_extension.cpp:84]: This needs to be empty.SymInt now

**Review by albanD — APPROVED** [review]
SGTM thanks

**Review by malfet — APPROVED** [review]
